### PR TITLE
Offering an alternative to the deprecated apt-key tool

### LINKED ISSUE #8298: debian repository keys
It's better to put .asc or .gpg file into `/etc/apt/trusted.gpg.d``` like
```
sudo wget -O /etc/apt/trusted.gpg.d/docker.asc https://download.docker.com/linux/debian/gpg
```

instead of apt-key add
```
curl -s /etc/apt/trusted.gpg.d/docker.asc | sudo apt-key add -
```

man APT-KEY(8):
>Note: Instead of using this command a keyring should be placed directly in the /etc/apt/trusted.gpg.d/ directory with a descriptive name and either "gpg" or "asc" as file extension.

### LINKED ISSUE #8341: Piping arbitrary input into `apt-key add` seems unsafe - why not use a command that may only add a single key, and only AFTER fingerprint is verified?
### Problem description

The documentation here https://docs.docker.com/install/linux/docker-ce/ubuntu/ / https://docs.docker.com/install/linux/docker-ee/ubuntu/  suggests the following command to add the docker key:

```shell
curl -fsSL "${DOCKER_EE_URL}/ubuntu/gpg" | sudo apt-key add -
```

However,

    1. it seems somewhat unobvious that this may not do a DoS by adding 999 keys

    2. when the next verification step fails, I am still stuck with a bad key

Isn't there a better targeted one that downloads and only adds **after** verifying it? Similar to the following one **but a verifying alternative:**

`apt-key adv --keyserver hkp://p80.pool.sks-keyservers.net:80 --recv-keys 58118E89F3A912897C070ADBF76221572C52609D`

(since while this has the long key id and therefore could possibly verify, it doesn't appear to be doing that in practice either:
```
    adv
           Pass advanced options to gpg. With adv --recv-key you can e.g. download key from
           keyservers directly into the the trusted set of keys. Note that there are no checks
           performed, so it is easy to completely undermine the apt-secure(8) infrastructure if
           used without care.
```
)

### Problem location

- I saw a problem on the following URL: https://docs.docker.com/install/linux/docker-ee/ubuntu/ 

### Suggestions for a fix

See above, it should be a command where the key fingerprint is checked *before* it is added system-wide.

### LINKED ISSUE #8725: why sudo the call to apt-key fingerprint?
File: [install/linux/docker-ce/ubuntu.md](https://docs.docker.com/install/linux/docker-ce/ubuntu/) Does apt-key fingerprint require sudo on some systems?

### LINKED ISSUE #10347: [RFI] Adding the docker repo's GPG key to the system.
### Problem description
- `apt-key(8)` <blockquote>
Note that if usage of apt-key is desired the additional installation of the GNU Privacy Guard suite packaged in gnupg) is required.
For this reason alone the programmatic usage (especially in package maintainer scripts!) <b><i><u>is strongly discouraged</u></i></b>.
</blockquote>

- Current `apt` expects keys from packages to be put inside `/etc/apt/trusted.gpg.d`
  - binary keys ending in `.gpg`
  - ASCII-armored keys ending in `.asc`

### Problem location
File: https://docs.docker.com/install/linux/docker-ce/ubuntu/#set-up-the-repository
- [ubuntu.md#L114](https://github.com/docker/docker.github.io/blame/master/install/linux/docker-ce/ubuntu.md#L114) (and maybe other documentation also) instructs to download the [docker repo's gpg key](https://download.docker.com/linux/ubuntu/gpg), which is actually an **ASCII-armored** file, and pipe it into `apt-key` which is deprecated on modern systems.
- [ubuntu.md#L153-L156](https://github.com/docker/docker.github.io/blame/master/install/linux/docker-ce/ubuntu.md#L153-L156) and similar lines.

### Suggestions for a fix
1. The `apt-key` command should **not be used** in the instructions.
2. The **recommended** command to retrieve the gpg-key should be changed into:\
   (As a single command-line)\
   <kbd>sudo curl -LR https://download.docker.com/linux/ubuntu/gpg -o <var>/etc/apt/trusted.gpg.d/docker.gpg.asc</var></kbd>
   - Notes:
     1. The `-fsS` flags to `curl` should not be used so the user has visual feedback and can retry in case of failures.
     2. The time-stamp of the source is used for the output file to aid in version checking.
     3. The output is put into a file under `/etc/apt/trusted.gpg.d` ending in `.asc` to indicate an _ASCII-armored key_ to `apt`.
3. The **recommended** command to check that the key is properly installed should be changed into:\
   <kbd>apt-key list docker</kbd>
4. The **recommended** command(s) to set up the _"stable"_ repository should be changed into:
	```console
	> printf "%s %s %s %s %s\n" \
		"deb" \
		"[arch=amd64]" \
		"https://download.docker.com/linux/ubuntu" \
		"$(lsb_release -cs)" \
		"stable" \
		> /tmp/docker.list
	> sudo install \
		--mode=a=r,u+w \
		/tmp/docker.list \
		/etc/apt/sources.list.d
	> rm /tmp/docker.list
	```

### Related reports
#1218, #8298, #9373, #10285, and maybe #10333?

### LINKED ISSUE #11625: Use trusted.gpg.d instead of apt-key
### Problem description

One of the steps at https://docs.docker.com/engine/install/ubuntu/#install-using-the-repository is to add Docker’s official GPG key like this:
```
$ curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -
```
However, the last part generates the following warning (at least on Ubuntu 20.10):
> Warning: apt-key is deprecated. Manage keyring files in trusted.gpg.d instead (see apt-key(8)).

Maybe it is time to use `trusted.gpg.d` as suggested by the warning?

### LINKED ISSUE #11776: Simple setup does not work on Ubuntu 20.10 as described.
File: [engine/install/ubuntu.md](https://docs.docker.com/engine/install/ubuntu/)

I'm not sure what exactly is going on, but I just created a brand new VM on azure and followed the documentation on the page.

Doing this I ran into the following a warning first:
```
....@Server:/etc/apt/trusted.gpg.d$ curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -
Warning: apt-key is deprecated. Manage keyring files in trusted.gpg.d instead (see apt-key(8)).
OK
```

Then I ran this command:
```
*****@Server:/etc/apt/trusted.gpg.d$ sudo add-apt-repository \
>    "deb [arch=amd64] https://download.docker.com/linux/ubuntu \
>    $(lsb_release -cs) \
>    stable"
Repository: 'deb [arch=amd64] https://download.docker.com/linux/ubuntu groovy stable'
Description:
Archive for codename: groovy components: stable
More info: https://download.docker.com/linux/ubuntu
Adding repository.
Press [ENTER] to continue or Ctrl-c to cancel.
Found existing deb entry in /etc/apt/sources.list.d/archive_uri-https_download_docker_com_linux_ubuntu-groovy.list
Adding deb entry to /etc/apt/sources.list.d/archive_uri-https_download_docker_com_linux_ubuntu-groovy.list
Found existing deb-src entry in /etc/apt/sources.list.d/archive_uri-https_download_docker_com_linux_ubuntu-groovy.list
Adding disabled deb-src entry to /etc/apt/sources.list.d/archive_uri-https_download_docker_com_linux_ubuntu-groovy.list
Hit:1 https://download.docker.com/linux/ubuntu groovy InRelease
Hit:2 http://azure.archive.ubuntu.com/ubuntu groovy InRelease
Hit:3 http://azure.archive.ubuntu.com/ubuntu groovy-updates InRelease
Hit:4 http://azure.archive.ubuntu.com/ubuntu groovy-backports InRelease
Hit:5 http://security.ubuntu.com/ubuntu groovy-security InRelease
Reading package lists... Done
```

Then:
```
****@Server:/etc/apt/trusted.gpg.d$ sudo apt-get update
Hit:1 http://azure.archive.ubuntu.com/ubuntu groovy InRelease
Hit:2 http://azure.archive.ubuntu.com/ubuntu groovy-updates InRelease
Hit:3 http://azure.archive.ubuntu.com/ubuntu groovy-backports InRelease
Hit:4 https://download.docker.com/linux/ubuntu groovy InRelease
Hit:5 http://security.ubuntu.com/ubuntu groovy-security InRelease
Reading package lists... Done
```

Then:
```
****@Server:/etc/apt/trusted.gpg.d$ sudo apt-get install docker-ce docker-ce-cli containerd.io
Reading package lists... Done
Building dependency tree
Reading state information... Done
Package docker-ce is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'docker-ce' has no installation candidate
E: Unable to locate package docker-ce-cli
```

I mean this doesn't really feel like I'm doing anything strange, however it does not work as you can see above. The package docker-ce does not exist.

### LINKED ISSUE #11851: Apt-key is deprecated in 20.20
File: [engine/install/ubuntu.md](https://docs.docker.com/engine/install/ubuntu/)

```
curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -
```

Should be replaced with something like

```
wget https://download.docker.com/linux/ubuntu/gpg -O /etc/apt/trusted.gpg.d/docker.gpg
```

### LINKED ISSUE #11915: Apt-Key is Depreciated
File: [engine/install/debian.md](https://docs.docker.com/engine/install/debian/)

### PR DESCRIPTION
**Pull request by denis-roy**
### Proposed changes

As of Debian 10 / Ubuntu 20.10, apt-key is deprecated and will not be available after Debian 11 / Ubuntu 22.04

Although adding keys directly to `/etc/apt/trusted.gpg.d`/ is suggested by apt-key deprecation message, as per [Debian Wiki](https://wiki.debian.org/DebianRepository/UseThirdParty) GPG keys for third party repositories should be added to `/usr/share/keyrings` and referenced with the `signed-by` option in the source.list.d entry.

With the proposed changes, verifying the key by fingerprint is unneeded and was removed.

### Applicable versions

The proposed changes are targeted at Debian 9 (Stretch) / Ubuntu 16.04 (Xenial) and onward, which is corresponding to Docker's own requirements described in the respective *OS requirements* sections ([Debian.md](https://github.com/denis-roy/docker.github.io/blob/patch-1/engine/install/debian.md#os-requirements) & [Ubuntu.md](https://github.com/denis-roy/docker.github.io/blob/patch-1/engine/install/ubuntu.md#os-requirements))

### Side note

Providing a binary .gpg key instead of the current ASCII Armored one might help shorten the lengthy command by removing the need for the ` | gpg --dearmor ` bit.

### Related issues & PRs

Closes #11625, closes #11851, closes #11915, closes #10778, closes #10545, closes #10505, closes #10010, closes #9373, closes #11974, closes #10347, closes #8298, closes #8725, closes #8341, closes #11776, closes #12313

### CODE REVIEW

**Comment by m-tmatma** [comment]
@denis-roy 

I tried the following commands on ubuntu 20.10 while referencing this PR.

```
sudo apt-get update  -y
sudo apt-get install -y \
    apt-transport-https \
    ca-certificates \
    curl \
    gnupg-agent \
    software-properties-common

curl -fsSL https://download.docker.com/linux/ubuntu/gpg | gpg --dearmor | sudo tee /usr/share/keyrings/docker-ce-archive-keyring.gpg > /dev/null

sudo add-apt-repository \
   "deb [arch=amd64 signed-by=/usr/share/keyrings/docker-ce-archive-keyring.gpg] https://download.docker.com/linux/ubuntu \
   $(lsb_release -cs) \
   stable"
```

But `add-apt-repository` failed with the following error.

```
`Unable to handle repository shortcut 'deb [arch=amd64 signed-by=/usr/share/keyrings/docker-ce-archive-keyring.gpg] https://download.docker.com/linux/ubuntu    groovy    stable'
```

Did I make a mistake?

**Comment by denis-roy** [comment]
> Did I make a mistake?

No @m-tmatma you did not. But I did and I'm sorry I did! :)

I clearly pasted from the wrong source since the intended change was to replace `add-apt-repository` by creating an entry in `/etc/apt/source.list.d/` directly, as the latest commit reflects

Thank you for pointing it out! :v:

**Comment by m-tmatma** [comment]
@denis-roy 

it worked.

I used the following command on clean-installed ubuntu 20.10 of VMWARE 16 player.

```
sudo apt-get update && sudo apt-get upgrade -y
sudo apt install -y open-vm-tools-desktop
```

```
#!/bin/sh

sudo apt-get update

sudo apt-get install -y \
    apt-transport-https \
    ca-certificates \
    curl \
    gnupg-agent \
    software-properties-common

curl -fsSL https://download.docker.com/linux/ubuntu/gpg | gpg --dearmor | sudo tee /usr/share/keyrings/docker-ce-archive-keyring.gpg > /dev/null
echo \
  "deb [arch=amd64 signed-by=/usr/share/keyrings/docker-ce-archive-keyring.gpg] https://download.docker.com/linux/ubuntu \
  $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker-ce.list > /dev/null
sudo apt-get update
sudo apt-get install -y docker-ce docker-ce-cli containerd.io
```

```
$ docker --version
Docker version 20.10.1, build 831ebea
```

**Comment by m-tmatma** [comment]
@denis-roy 

You can remove the installation of `software-properties-common`  because you don't use `add-apt-repository`.

I re-confirmed these commands on clean-installed Ubuntu 20.10.

```
#!/bin/sh

sudo apt-get update

sudo apt-get install -y \
    apt-transport-https \
    ca-certificates \
    curl \
    gnupg-agent

curl -fsSL https://download.docker.com/linux/ubuntu/gpg | gpg --dearmor | sudo tee /usr/share/keyrings/docker-ce-archive-keyring.gpg > /dev/null
echo \
  "deb [arch=amd64 signed-by=/usr/share/keyrings/docker-ce-archive-keyring.gpg] https://download.docker.com/linux/ubuntu \
  $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker-ce.list > /dev/null
sudo apt-get update
sudo apt-get install -y docker-ce docker-ce-cli containerd.io
```

**Comment by denis-roy** [comment]
> @denis-roy
> 
> You can remove the installation of `software-properties-common` because you don't use `add-apt-repository`.

True that! Changes made, thanks again :) :v:

**Comment by usha-mandya** [comment]
@denis-roy Thanks for the PR.

@thaJeztah Could you PTAL?

File: engine/install/debian.md [hunk 1]
```diff
@@ -84,27 +84,14 @@ from the repository.
...
-    $ curl -fsSL {{ download-url-base }}/gpg | sudo apt-key add -
-    ```
-
-    Verify that you now have the key with the fingerprint
-    `9DC8 5822 9FC7 DD38 854A  E2D8 8D81 803C 0EBF CD88`, by searching for the
-    last 8 characters of the fingerprint.
-
-    ```bash
-    $ sudo apt-key fingerprint 0EBFCD88
+    $ curl -fsSL {{ download-url-base }}/gpg | gpg --dearmor | sudo tee /usr/share/keyrings/docker-ce-archive-keyring.gpg > /dev/null
```
Comment (thaJeztah) [inline engine/install/debian.md:93]: Wondering if we need some step after this to (manually) verify the fingerprint

File: engine/install/ubuntu.md [hunk 2]
```diff
@@ -92,27 +92,13 @@ from the repository.
...
-    last 8 characters of the fingerprint.
-
-    ```bash
-    $ sudo apt-key fingerprint 0EBFCD88
-
-    pub   rsa4096 2017-02-22 [SCEA]
-          9DC8 5822 9FC7 DD38 854A  E2D8 8D81 803C 0EBF CD88
-    uid           [ unknown] Docker Release (CE deb) <docker@docker.com>
-    sub   rsa4096 2017-02-22 [S]
+    $ curl -fsSL {{ download-url-base }}/gpg | gpg --dearmor | sudo tee /usr/share/keyrings/docker-ce-archive-keyring.gpg > /dev/null
```
Comment (thaJeztah) [inline engine/install/ubuntu.md:101]: Same here

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks! Left a comment

Also would love @tianon to have a look at this 🤗

File: engine/install/debian.md (hunk 1 above)
Comment (tianon) [reply engine/install/debian.md:93]: If we add `.asc` to the filename, we don't need to `--dearmor` :sweat_smile:

Although that makes it slightly more difficult to add a verification step. :thinking:

(I personally use `gpg --keyserver ha.pool.sks-keyservers.net --recv-keys 9DC858229FC7DD38854AE2D88D81803C0EBFCD88` to fetch this key, which directly verifies it, but it's then much less reliable in fetching, ala https://github.com/docker-library/faq#openpgp--gnupg-keys-and-verification)

Maybe it would be enough to list the fingerprint here and leave it as an exercise for the reader to determine how to verify it?  We could provide the hint of `gpg --import` + `gpg --fingerprint` which are the two commands that will help them do so.

I also wonder whether `/usr/share/keyrings/...` is the "right" place for us to suggest users be downloading this file.  I don't know that there's really an "official" place for sideloaded keyrings like this (I personally use `/etc/apt/tianon.gpg.d` :sweat_smile:), but `/usr/share` is traditionally supposed to be filled by the package manager (with `/usr/local/share` being the "user supplied" variant), so perhaps we instead suggest users create `/usr/local/share/keyrings` and put it there?

(Other than those, I'm a big fan of this PR!  :smile: :metal:)

Comment (thaJeztah) [reply engine/install/debian.md:93]: Where does `apt-key add` store it? Does that specify some "canonical" path where to store the information?

Comment (tianon) [reply engine/install/debian.md:93]: It stores it inside `/etc/apt/trusted.gpg` - there exists an `/etc/apt/trusted.gpg.d/` that is an appropriate place for these things, but any keys added in either of those are trusted for *all* repo entries, which is why this added the additional `signed-by=` bit to the repo line (so that the key is only trusted for that one repo instead, which is much safer and the recommended solution by the APT and Debian GnuPG maintainers).

Comment (denis-roy) [reply engine/install/debian.md:93]: > I also wonder whether `/usr/share/keyrings/...` is the "right" place

@tianon, it's where [Debian's own Wiki](https://wiki.debian.org/DebianRepository/UseThirdParty#OpenPGP_Key_distribution) suggests we put it :shrug: :)


Quote (emphasis theirs):
>> The key MUST be downloaded over a secure mechanism like HTTPS to a location only writable by root, which SHOULD be /usr/share/keyrings.

@thaJeztah, I believe `apt-key` appends them to the `/etc/apt/trusted.gpg` file

Comment (thaJeztah) [reply engine/install/debian.md:93]: > I believe apt-key appends them to the /etc/apt/trusted.gpg file

Ah, thx

Is the name of the file important, or would any name do?  Looking if `docker-ce-archive-keyring.gpg` is a good fit, or if we should use something else (FWIW, now that "Docker EE" is no longer a thing, we are toning down the "CE" moniker a bit; we won't be renaming packages as we know that's quite disruptive, but perhaps something more generic could work for the filename)

Comment (denis-roy) [reply engine/install/debian.md:93]: > Is the name of the file important, or would any name do?

Any name will do both for the `.gpg` keyring and the `.list` file and they can be different from one another, so we can future-proof them with generic filenames as you wish :)

**Comment by MichaIng** [comment]
I should have searched for similar fixes, made one as well, which the aim first to remove `add-apt-repository`, which causes issues on Raspbian and probably other Debian-based distros, then replaced `apt-key` in that turn: #12331

Now things are doubled 😄. Good about this one is that it aligns the Ubuntu instructions as well. And I like the `signed-by=` method to have APT verifying the key signing status, to remove the last "apt-key" call as well.

What could be adopted here from my PR is:
- Using `gpg -o /path/to/key.gpg` argument to avoid another `tee` in the pipe.
- Replace `gnupg-agent` with `gnupg`. The first does not pull in the `gpg` command, which the seconds does. On Debian Buster and above, there is a `gpg` package, on Stretch it's a meta package for pulling in `gnupg`. However with Stretch backports enabled, `apt-get install gpg` tries to pull `gpg` from backports, with then missing backports dependencies. It's generally a good idea to pull the whole `gnupg` suite to assure that everything can deal with this keys.

The remaining changes in my PR are to reformat instructions so that Raspbian is supported with all methods. I propose that this PR here is merged first, and I can merge and resolve conflicts with mine afterwards.

File: engine/install/debian.md (hunk 1 above)
Comment (MichaIng) [reply engine/install/debian.md:93]: `apt-key add` appends keys to `/etc/apt/trusted.gpg`, which is not used anymore by most modern distro versions. Usually keys are added to `/etc/apt/trusted.gpg.d/name.gpg`, which is queried by APT automatically. So I suggest to use `/etc/apt/trusted.gpg.d/docker.gpg` to simplify and align it with how Debian and Ubuntu do it on current stable and testing versions.

To avoid the `sudo tee` pipe, `gpg` actually has an output argument, so the following would be simpler:
```
curl -fsSL {{ download-url-base }}/gpg | sudo gpg --dearmor -o /etc/apt/trusted.gpg.d/docker.gpg --yes
```
`--yes` allows to overwrite an existing file, e.g. in case of an update. But probably this is better skipped to not risk overwriting other keys and in case let the admin decide about this interactively.

**Comment by denis-roy** [comment]
@MichaIng - Thanks for your input

> * Using `gpg -o /path/to/key.gpg` argument to avoid another `tee` in the pipe.

Definitely :+1: 
 
> * Replace `gnupg-agent` with `gnupg`. The first does not pull in the `gpg` command, which the seconds does. On Debian Buster and above, there is a `gpg` package, on Stretch it's a meta package for pulling in `gnupg`. However with Stretch backports enabled, `apt-get install gpg` tries to pull `gpg` from backports, with then missing backports dependencies. It's generally a good idea to pull the whole `gnupg` suite to assure that everything can deal with this keys.

 :+1: here as well 

> * Write to key to `/etc/apt/trusted.gpg.d/something.gpg` to have it used by APT automatically and align with the way how current stable and testing Debian and Ubuntu handle it with their own keys.

Part of me agrees and the other part of me still want to leans towards Debian Wiki's `/usr/share/keyrings` suggestion which I have started to use and quite like so will leave this one up for @tianon & @thaJeztah to decide :game_die: 

Thanks again :)

**Comment by MichaIng** [comment]
**TL;DR: While I find /etc/apt/trusted.gpg.d vs /usr(/local)/share/keyrings an interesting topic, since both ways work just fine, it's not something that must block this PR from being merged IMO 😉. It's perfectly fine as it is my end!**

The Debian keyring package btw ships the keys in both directories: /var/lib/dpkg/info/debian-archive-keyring.conffiles
However, debhelper packaging adds all files in /etc automatically to conffiles:
```
# cat /var/lib/dpkg/info/debian-archive-keyring.conffiles
/etc/apt/trusted.gpg.d/debian-archive-buster-automatic.gpg
/etc/apt/trusted.gpg.d/debian-archive-buster-security-automatic.gpg
/etc/apt/trusted.gpg.d/debian-archive-buster-stable.gpg
/etc/apt/trusted.gpg.d/debian-archive-jessie-automatic.gpg
/etc/apt/trusted.gpg.d/debian-archive-jessie-security-automatic.gpg
/etc/apt/trusted.gpg.d/debian-archive-jessie-stable.gpg
/etc/apt/trusted.gpg.d/debian-archive-stretch-automatic.gpg
/etc/apt/trusted.gpg.d/debian-archive-stretch-security-automatic.gpg
/etc/apt/trusted.gpg.d/debian-archive-stretch-stable.gpg
```
This means that those are not automatically updated with the package, but only if the admin interactively decides so during package upgrade or uses the `--force-confnew` dpkg option. Also when any of those is removed, it stays removed with a package upgrade, unless `--force-confmiss` is used. Since these files are essential for the purpose of the package, it hence makes sense to keep the original keys in a separate directory where it's assured that they are restored and updated together with the package, which then is reasonable for all keyring DEB packages, when assuming that debhelper is used and that no package shall forcefully override by admin removed or edited files. For an installer script however, is not executed as part of "apt upgrade" but only manually by the admin, these arguments do not fully apply.

I found this wiki, which suggests to use trusted.gpg.d: https://wiki.debian.org/SecureApt#Basic_concepts
The [Wiki page](https://wiki.debian.org/DebianRepository/UseThirdParty#OpenPGP_Key_distribution) you linked above is newer and contrary states:
> The key MUST NOT be placed in /etc/apt/trusted.gpg.d or loaded by apt-key add.
> ...
> The reason we point to a file instead of a fingerprint is that the latter forces the user to add the key to the global SecureApt trust anchor in /etc/apt/trusted.gpg.d, which would cause the system to accept signatures from the third-party keyholder on all other repositories configured on the system that don't have a signed-by option (including the official Debian repositories).

On that argument: But no other configured repository is able to ship a signature that satisfies this key, otherwise the whole signing effort would all be nonsense 😄. And since Debian ships it's keyrings in trusted.gpg.d, the other way round, Debian signatures from Debian are accepted for all other repositories, so I don't see why this rule should apply for 3rd party repositories but not for Debian itself. If one does not trust a third party repository, it must not be added, that's it, and that's the same for the Debian repository. To me it looks like this Wiki page was written from a very single-sided perspective where Debian is trusted and 3rd party repositories are not trusted, but that is naturally not a trustable perspective.
Or do I oversee something?

Interesting topic definitely, especially since we (DietPi OS) are using /etc/apt/trusted.gpg.d for third party repositories currently, and I'm now considering to switch to /usr/local/share/keyrings or similar to satisfy Debian's rules according to this newer Wiki, if those those turn out to be reasonable 🤔.

**Review by MichaIng — APPROVED** [review]

**Review by denis-roy — COMMENTED** [review]
Fixed by these commit:

- https://github.com/docker/docker.github.io/pull/11990/commits/11c3fb6b4517e9e4bfb18be93de8d70bbeadc813
- https://github.com/docker/docker.github.io/pull/11990/commits/694aaad93ec467821a608975cdc113720289c916

**Comment by MichaIng** [comment]
Btw: `gpg -o` has the other benefit, that it does not create/overwrite the file with an empty file when it fails, but similarly to `curl -o` the output file is only created/overwritten when the command succeeds. Also without `--yes`, `gpg` will ask for verification if the file exists already, which is IMO a good safety measure as well.

File: engine/install/debian.md (hunk 1 above)
Comment (tianon) [reply engine/install/debian.md]: Just for posterity, I clarified the text from https://wiki.debian.org/DebianRepository/UseThirdParty#OpenPGP_Key_distribution with some other Debian Developers and the consensus was that the intent there was likely that the user is creating a file which will almost immediately be overwritten (and thus then owned/managed) by a file provided by a `deriv-archive-keyring` package installed from that overlay, which is how most derivatives work and manage keyring updates in a similar way to how Debian itself does so.

That's not the case for Docker, but it's reasonably harmless to use the non-local path anyhow (the chance for conflicts on this filename is extremely low, and the minority of users who are dogmatic about the FHS will grumble as they read this and adjust the path for themselves anyhow :smile:). :+1:

**Review by tianon — APPROVED** [review]
LGTM

**Review by MichaIng — APPROVED** [review]

File: engine/install/ubuntu.md [hunk 3]
```diff
@@ -92,27 +92,13 @@ from the repository.
...
 
 2.  Add Docker's official GPG key:
 
     ```bash
-    $ curl -fsSL {{ download-url-base }}/gpg | sudo apt-key add -
-    ```
-
-    Verify that you now have the key with the fingerprint
-    <span><code>9DC8 5822 9FC7 DD38 854A&nbsp;&nbsp;E2D8 8D81 803C 0EBF CD88</code></span>, by searching for the
-    last 8 characters of the fingerprint.
```
Comment (thaJeztah) [inline engine/install/ubuntu.md:107]: I was looking at steps to show how to verify (will post a follow-up PR for review/consideration)

File: engine/install/ubuntu.md [hunk 4]
```diff
@@ -135,30 +121,27 @@ from the repository.
     <div id="x86_64_repo" class="tab-pane fade in active" markdown="1">
 
     ```bash
-    $ sudo add-apt-repository \
-       "deb [arch=amd64] {{ download-url-base }} \
-       $(lsb_release -cs) \
-       stable"
+    $ echo \
+      "deb [arch=amd64 signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] {{ download-url-base }} \
+      $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
```
Comment (thaJeztah) [inline engine/install/ubuntu.md:126]: not a blocker; mostly curious; what's the advantage of doing `echo | sudo tee` instead of `sudo echo > /etc/apt...` ? Is it so that `lsb_release -cs` is not executed as `root` ?

**Review by thaJeztah — APPROVED** [review]
Thank you! Sorry for the long delay; this one kept dropping down my list 😊 😓.

LGTM

File: engine/install/ubuntu.md (hunk 4 above)
Comment (MichaIng) [reply engine/install/ubuntu.md:126]: `sudo echo >` does not work, as only the echo command itself had root permissions, while the redirect/file write itself has not and hence would fail. So `echo | sudo tee` is a necessity to grant root permissions for the file write itself, not a security measure.

`lsb_release` is always executed as original user, regardless if `$(lsb_release -cs)` is part of the sudo'ed command or any other part of the pipe. Like variables, command substitutions are expanded before the resulting string is passed to the parent command. Hence `sudo parent "$(child)"` would execute `child` as original user and pass the resulting string as argument to `parent` and only that is executed as root.

But indeed it could be seen as security measure to grant root permissions only to the final file write, so that, if the other commands were maliciously manipulated, they cannot do more than affecting the string that is written to a defined file. For the `gpg` command above that would be an argument (where a `sudo` is missing, I see now 😃):
```
sudo gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg
```
vs
```
gpg --dearmor | sudo tee /usr/share/keyrings/docker-archive-keyring.gpg
```
But I mean `tee` could be manipulated as well, and considering the other benefits that `-o` has (not creating an empty file when it fails), I would keep it the way it is now.

File: engine/install/ubuntu.md [hunk 5]
```diff
@@ -92,27 +92,13 @@ from the repository.
...
-    last 8 characters of the fingerprint.
-
-    ```bash
-    $ sudo apt-key fingerprint 0EBFCD88
-
-    pub   rsa4096 2017-02-22 [SCEA]
-          9DC8 5822 9FC7 DD38 854A  E2D8 8D81 803C 0EBF CD88
-    uid           [ unknown] Docker Release (CE deb) <docker@docker.com>
-    sub   rsa4096 2017-02-22 [S]
+    $ curl -fsSL {{ download-url-base }}/gpg | gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg
```
Comment (MichaIng) [inline engine/install/ubuntu.md:101]: `sudo` is missing for `gpg`, compare with Debian instructions above.

File: engine/install/ubuntu.md (hunk 4 above)
Comment (MichaIng) [reply engine/install/ubuntu.md:126]: This seems to work:
```
# gpg --fingerprint --no-default-keyring --keyring /etc/apt/trusted.gpg.d/debian-archive-buster-stable.gpg
/etc/apt/trusted.gpg.d/debian-archive-buster-stable.gpg
-------------------------------------------------------
pub   rsa4096 2019-02-05 [SC] [expires: 2027-02-03]
      6D33 866E DD8F FA41 C014  3AED DCC9 EFBF 77E1 1517
uid           [ unknown] Debian Stable Release Key (10/buster) <debian-release@lists.debian.org>
```

Comment (thaJeztah) [reply engine/install/ubuntu.md:126]: > sudo echo > does not work, as only the echo command itself had root permissions, while the redirect/file write itself has not and hence would fail.

oh! you're right 🤦 

> This seems to work:

Thanks! I made some notes / possible approaches for checking the fingerprint (I recall there were some of those options that worked on the more "recent" versions of Debian/Ubuntu, but were not yet available on the older versions.

I'll dig up those notes, and will open a pull request later today

File: engine/install/ubuntu.md (hunk 5 above)
Comment (thaJeztah) [reply engine/install/ubuntu.md:101]: 👍 I noticed after merging, and did a follow-up PR to fix it; https://github.com/docker/docker.github.io/pull/12422

**Comment by denis-roy** [comment]
Thank you to all who participated 👍

**Comment by doolio** [comment]
@denis-roy has these changes since been removed from the official docs? It seems so.
